### ISSUE TITLE
SPDY is deprecated. Switch to HTTP/2.

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#7452]
Filing for tracking purposes.

See #7392 where the spdy package we use doesn't even exist any more. (And btw it has bugs anyway, possibly security bugs-- see the trophies here: https://github.com/dvyukov/go-fuzz)

Blocked on github.com/docker/spdystream switching to HTTP/2.

(Note: HTTP/2 is based on SPDY, with just a few differences-- hopefully this transition will be easy.)

@thockin @ncdc

**Comment by ncdc** [comment]
Waiting on HTTP/2 support for Go that includes low level client/server ability to create and handle streams (https://github.com/bradfitz/http2).

https://github.com/docker/spdystream/pull/53 pulls the upstream spdy code into spdystream.

**Comment by lavalamp** [comment]
Ah, after docker/spdystream#53 is merged, we can at least make godep happy again. :)

**Comment by ncdc** [comment]
@lavalamp spdystream PR is merged, FYI

**Comment by ncdc** [comment]
@thockin @lavalamp it looks like the Go HTTP/2 issue was just closed (https://github.com/golang/go/issues/6891#issuecomment-161070230). I glanced quickly through the code, but I don't see a way to hijack the connection and create our own streams, like we do with spdystream. Hopefully we can get this functionality in there at some point...

**Comment by ncdc** [comment]
cc @smarterclayton

**Comment by ncdc** [comment]
I created https://github.com/golang/go/issues/13444 as an RFE for end-user stream support

**Comment by ddysher** [comment]
What's the plan for the deprecation?  We are using nginx in front of apiserver which drops support for SPDY since version 1.9.5.  Our workaround is to use an older nginx, but would like to see http/2 support go upstream :)

**Comment by ncdc** [comment]
I prototyped this with a prerelease version of go1.6. It worked somewhat, but there were some rough spots (probably stuff I was doing incorrectly). We can't move to http/2 until go1.6 at the earliest, and even then we'll have to do R&D to see if it's technically feasible. I had been hoping that the http/2 support in go1.6 would allow me to create streams just like we can do in spdystream, but that isn't exposed via the API. For my prototype, I ended up writing a simple multiplexer just like @smarterclayton wrote for the websocket support we have in Kubernetes now.

**Comment by ddysher** [comment]
Thanks for the heads up. Looking at the linked comments, there is no intention to add the support in go1.6, so I guess it's probably still months away.

**Comment by JensRantil** [comment]
> We can't move to http/2 until go1.6 at the earliest

Golang 1.7 is out now. Does that mean we can make progress on this issue now? Anything else stopping us?

**Comment by smarterclayton** [comment]
A usable HTTP/2 framer in Go that allows us hijack control and the ability
to stand up the connection.  Last I saw we were still blocked without
explicitly forking the Go implementation.

In the meantime, websockets continue to be an option (with portforward
being the last bit remaining).

On Oct 29, 2016, at 2:22 PM, Jens Rantil <notifications@github.com> wrote:

We can't move to http/2 until go1.6 at the earliest

Golang 1.7 is out now. Does that mean we can make progress on this issue
now? Anything else stopping us?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/7452#issuecomment-257107405>,
or mute the thread

**Comment by ncdc** [comment]
We should be able to use HTTP/2 without any changes to the Go implementation. But we would have to write our own code to do multiplexing of multiple data streams over a single request/response body pair, which we'd rather not do if we can avoid it.

**Comment by bradfitz** [comment]
@ncdc, if it helps, https://godoc.org/golang.org/x/build/revdial is used by the Go build system to multiplex multiple net.Conns and a net.Listener all over a single net.Conn.

**Comment by fraenkel** [comment]
@bradfitz Thanks for the pointer, but it doesn't fix the underlying issue. While SPDY is being used, it is not following HTTP request/response semantics. A logical connection is created and then data is sent in either direction as needed. It would be equivalent to the client sending an HTTP request and the server doing a Push whenever.

**Comment by ncdc** [comment]
@fraenkel I don't think there's any technical limitation in the facilities provided by the Go HTTP/2 implementation. As I wrote above, we would just need to write our own muxing/framing code, which we'd rather avoid (or find a library that can do it?).

**Comment by fraenkel** [comment]
@ncdc It might be easier to just build on top of gRPC which provides bi-directional streaming or copy what they have done.

**Comment by smarterclayton** [comment]
There's definitely benefit to implementing a gRPC endpoint vs continuing to
have to maintain a franken-protocol.  That is tied up with the CRI changes
though.

On Mon, Nov 28, 2016 at 11:25 AM, Michael Fraenkel <notifications@github.com

**Comment by luxas** [comment]
Ping, any update on this?

**Comment by wlan0** [comment]
We, at Rancher Labs are also interested in this change.

**Comment by smarterclayton** [comment]
We can actually just move to web sockets in the near term.  The server has supported it for a while and HOL blocking is not an issue

**Comment by smarterclayton** [comment]
Tracking #48633 to move kubectl to web sockets.

**Comment by smarterclayton** [comment]
General proposal:

Mark SPDY deprecated in 1.8.  Clients (https://github.com/kubernetes/kubernetes/issues/48633) switch to using websockets in 1.8/1.9/1.10.  We stop adding new streaming support for SPDY.

**Comment by luxas** [comment]
/remove-lifecycle stale
/lifecycle frozen

Assuming @smarterclayton is still working on https://github.com/kubernetes/kubernetes/pull/50428

**Comment by josdotso** [comment]
Is there any status update on this issue?

**Comment by mfranczy** [comment]
Please ignore my reference, it has nothing to do with that issue, if I could I would remove it.

**Comment by ensonic** [comment]
We run a http/2 reverse proxy to expose a on-prem api-server. This issues breaks kubectl exec and thereby also tools like helm.

**Comment by dims** [comment]
long-term-issue (note-to-self)

**Comment by mikedanese** [comment]
I've started working on streaming over h2 here:

https://github.com/kubernetes/kubernetes/compare/master...mikedanese:h2?expand=1

I hope to make some progress in 1.14

**Comment by mikedanese** [comment]
I'd like to see this happen in two parts:

1. support all streaming codecs over http/2.
2. deprecate spdy streaming codecs.

I think (1) is probably the biggest win.

**Comment by mikedanese** [comment]
I have a rough [prototype](https://github.com/kubernetes/kubernetes/compare/master...mikedanese:h2stream?expand=1) of (1) working e2e from kubectl to CRI. Migration will be tricky. I haven't thought of a way for clients to discover h2 streaming capabilities since the entire stack (kubectl -> apiserver -> kubelet -> cri) needs to support streaming over h2. With a significantly more complex implementation, the upgrade aware proxy could translate between clients and servers speaking different protocols. If that were the case, we'd likely only need single hop discovery to transition smoothly.

**Comment by mikedanese** [comment]
https://github.com/kubernetes/kubernetes/issues/71411 is an example of the dangers of dropping an http connection to TCP. There's a category of bugs to be avoided but this one was particularly insidious.

**Comment by scult** [comment]
Any update on this?

**Comment by mikedanese** [comment]
We are blocked on support for https://tools.ietf.org/html/rfc8441 in x/net/http2 if we don't want to implement something non-standard again.

**Comment by jayunit100** [comment]
Is it possible for the API server to support HTTP2 for watches, while still using HTTP1 for everything else?

**Comment by jmillikin-stripe** [comment]
Has there been any progress on this since May? The use of SPDY for `kubectl exec` continues to be a pain point because it's no longer supported by common network proxies (nginx, haproxy, envoy, and so on).

Naively, I would expect something like interactive sessions to be implemented on top of gRPC or another thin layer on the normal HTTP/2 framing. Websockets, CONNECT, and other ways to get a raw byte stream have a difficult security model and need much more support in intermediate software compared to plain HTTP/2.

**Comment by jfrabaute** [comment]
Hi,

I have a setup where proxying `kubectl exec` is working, using openresty.
It should also work with nginx as openresty is based on nginx.

The relevant config in the `location` block to have `kubectl exec` proxied correctly is to add:

```
              proxy_set_header Upgrade $http_upgrade;
              proxy_set_header Connection "Upgrade";
```

this is the `listen` line in the `server` block:
```
listen 443 http2 ssl
```
The proxied kube-apiserver is running `1.14.3-gke.11` (so, k8s 1.14 with gke modified version).

**Comment by brooksmtownsend** [comment]
Hello all, any progress on this issue late 2019 / early 2020? We are using an AWS ALB for the rest of our application, but since AWS ALBs do not support the SPDY protocol we have to use a different load balancer solution just for the k8s API.

**Comment by elixxx** [comment]
Hey  @brooksmtownsend i am sure switching to AWS [NetworkLoadBalancer](https://aws.amazon.com/de/blogs/aws/new-network-load-balancer-effortless-scaling-to-millions-of-requests-per-second/) is the only solution at the moment. 
I tried for days to get it running, ALB and ELB don't support SPDY and i can't imagine AWS will add support. So please safe time and use NLB.

**Comment by vespian** [comment]
Is there any status update on this issue?

**Comment by librannk** [comment]
any update on this ??

**Comment by landorg** [comment]
> You have a proxy of apiserver over nginx? Is it possible to do the same with a nginx Ingress resource?

... or kong? as it's based on openresty somehow.

**Comment by adisky** [comment]
/remove lifecycle-frozen
Some work is going on here https://github.com/kubernetes/kubernetes/issues/89163

**Comment by wgahnagl** [comment]
still relevant! known tech debt but we don't have anyone working on fixing it :(

**Comment by arianitu** [comment]
Just want to mention that nginx does not support HTTP/2 when proxying. Are you guys sure you want to use HTTP/2?

Specifically:

client (http2) -> nginx -> (http2 does not work here) kube-api-server

We are running into essentially the same issue with our API (picking a good bidirectional streaming protocol) and we're trying to pick a good option that works well with proxying servers like nginx. We implemented HTTP/2 initially and then realized, oops, can't use it with nginx so now we are abandoning the implementation.

We are looking at replacing HTTP/2 with WebSockets which seem to work well with nginx.

**Comment by sfxworks** [comment]
I believe that would work out of the box with cloudflare

**Comment by slandelle** [comment]
> some calls (exec, attach) ... and it doesn't need to go through a proxy

@lavalamp Why is that?

**Comment by danopia** [comment]
I'm a user of Kubernetes so the below is my own viewpoint as an API consumer, not a Kubernetes contributor.

> > some calls (exec, attach) ... and it doesn't need to go through a proxy
> 
> Why is that?

When the API server makes HTTP calls to a particular kubelet, the connection is generally made without an HTTP proxy in the way; the kubelets each have their own TLS certificates and the apiserver checks this encryption, so installing an HTTP proxy in between would introduce MITM-like security concerns.

This connectivity of course may need to be relayed e.g. across the internet and this is done at the TCP level. For example GKE runs the apiserver->kubelet traffic through long-lived SSH tunnels, and the new system to replace those SSH tunnels, [Konnectivity](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/), is also a TCP-level proxy.

---

The client -> apiserver traffic is definitely more important for compatibility and this is where Websockets are already offered. Unfortunately the websocket interop isn't perfect:

1. The `new WebSocket()` constructor in browsers don't allow attaching an `Authorization` header, so token auth isn't possible from an actual web client.
1. Each WebSocket is historically its own TCP connection, so `kubectl port-forward` functionality would need to open a new TCP connection to the API Server for each new tunnel client if it were using WebSockets. RFC 8441 addresses this, but similarly, [nginx seems to have no plans to implement websocket over HTTP2](https://trac.nginx.org/nginx/ticket/1992).

I've been working on a WebSocket client to Kubernetes and you can get things done with it. If the websocket subprotocol used by Kubernetes was extended to support functions like dynamically opening and closing port-forward streams, I could see it replacing SPDY. I just don't know if creating a more special-cased subprotocol in Websocket would be worth it vs. leveraging the http2 standard.

**Comment by lavalamp** [comment]
Thanks for the websocket feedback. I don't know if anyone is working on this right now, though. It may be a good idea to file a separate issue about that for visibility.

**Comment by pires** [comment]
@danopia doesn't it make sense for the `Authorization` header go side-by-side with an `Upgrade` header in the same HTTP request to kube-apiserver and then, provided the requestor is authorized to _proxy_, upgrade to Websocket (not needing the Authorization header)?

**Comment by aeriksson** [comment]
> The new WebSocket() constructor in browsers don't allow attaching an Authorization header, so token auth isn't possible from an actual web client.

FYI it seems like there's a way to do this using the websocket subprotocol header: https://github.com/kubernetes/kubernetes/pull/47740

**Comment by danopia** [comment]
> FYI it seems like there's a way to do this using the websocket subprotocol header: #47740

That hack looks questionable at best, despicable at worst, and I'm sure not at all documented.

I've verified that the hack works for my usecase. Thanks for the link.

**Comment by maradwan** [comment]
I solved the issue, see my article 
https://dev.to/aws-builders/kubectl-execport-forward-with-aws-alb-and-nginx-ingress-controller-4fn0

**Comment by anders-swanson** [comment]
Hi, is this issue actively being worked on?

**Comment by lavalamp** [comment]
No --

We know the user facing path should be swapped out for websockets, but I don't know if there is any consensus on what should happen for the apiserver -> kubelet path.

**Comment by GreenCappuccino** [comment]
Given that https://github.com/kubernetes/enhancements/pull/3401 was closed, noting issues with implementing an application-layer multiplexing system for a websocket stream.

I'm really unsure as to how well this is ready for the k8s ecosystem, but it might be worth keeping an eye on [WebTransport](https://w3c.github.io/webtransport/#introduction), as it seems to natively support many of the desired features missing from WebSockets (multiplexing, bidirectional streams). However, afaik it's still being standardized, and depends on HTTP/3 (QUIC), which is (newer?) to k8s.

**Comment by dims** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/89163

**Comment by lavalamp** [comment]
BTW, the title of this issue scans to "Rudolph the red nosed reindeer".

**Comment by sftim** [comment]
Although WebTransport is not yet standardized, we do have both WebSocket and WebRTC Datachannels as (fairly) stable things.

WebSockets over HTTP/3 over QUIC is [standardized](https://datatracker.ietf.org/doc/html/rfc9220.html) albeit with few browser implementations. That could nonetheless work fine for communication between code we write (eg `kubectl` talking to the Kubernetes API server), with a fallback to an earlier HTTP version and TCP.

**Comment by aojea** [comment]
I'm going to close this, since KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets moved to websockets that is widely available I don't think we are going to do another protocol change

/close
